proxy/coco/src/state.rs

```diff
@@ -0,0 +1,733 @@
+            .collect::<Vec<_>>();
+        project_names.sort();
+
+        assert_eq!(
+            project_names,
+            vec!["Monadic", "monokel", "open source coin", "radicle"]
+        );
+
+        assert!(!project_names.contains(&fakie.name().to_string()));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn list_users() -> Result<(), Error> {
+        let tmp_dir = tempfile::tempdir().expect("failed to create temdir");
+        let key = SecretKey::new();
+        let signer = signer::BoxedSigner::from(key.clone());
+        let config = config::default(key, tmp_dir.path())?;
+        let (api, _run_loop) = config.try_into_peer().await?.accept()?;
+        let state = State::new(api, signer.clone());
+
+        let cloudhead = state.init_user(&signer, "cloudhead")?;
+        let _cloudhead = super::verify_user(cloudhead)?;
+        let kalt = state.init_user(&signer, "kalt")?;
+        let _kalt = super::verify_user(kalt)?;
+
+        let users = state.list_users()?;
+        let mut user_handles = users
+            .into_iter()
+            .map(|user| user.name().to_string())
+            .collect::<Vec<_>>();
+        user_handles.sort();
+
+        assert_eq!(user_handles, vec!["cloudhead", "kalt"],);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn create_with_existing_remote_with_reset() -> Result<(), Error> {
+        let tmp_dir = tempfile::tempdir().expect("failed to create tempdir");
+        let repo_path = tmp_dir.path().join("radicle");
+        let key = SecretKey::new();
+        let signer = signer::BoxedSigner::from(key.clone());
+        let config = config::default(key, tmp_dir.path())?;
+        let (api, _run_loop) = config.try_into_peer().await?.accept()?;
+        let state = State::new(api, signer.clone());
+
+        let kalt = state.init_owner(&signer, "kalt")?;
+
+        let fakie = state.init_project(&signer, &kalt, &fakie_project(repo_path.clone()))?;
+
+        assert!(repo_path.join(fakie.name()).exists());
+
+        // Simulate resetting the monorepo
+        let tmp_dir = tempfile::tempdir().expect("failed to create tempdir");
+        let key = SecretKey::new();
+        let signer = signer::BoxedSigner::from(key.clone());
+        let config = config::default(key, tmp_dir.path())?;
+        let (api, _run_loop) = config.try_into_peer().await?.accept()?;
+        let state = State::new(api, signer.clone());
+
+        // Create fakie project from the existing directory above.
+        let kalt = state.init_owner(&signer, "kalt")?;
+        let fakie =
+            state.init_project(&signer, &kalt, &fakie_project(repo_path).into_existing())?;
+
+        // Attempt to initialise a browser to ensure we can look at branches in the project
+        let _stats = state.with_browser(&fakie.urn(), |browser| Ok(browser.get_stats()?))?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn create_with_existing_remote() -> Result<(), Error> {
```

Comment:
I would highly benefit from having tests such as this one with a simple comment briefly explaining what it's testing. I'd need to read the whole thing to understand what it is trying to create and what it means to create _it_with an existing remote.

Reply:
What would make it easier for you to understand?